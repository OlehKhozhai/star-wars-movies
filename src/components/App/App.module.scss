$mobile: 480px;
$tablet: 768px;
$desktop: 1440px;

.root {
  display: flex;
  flex-flow: row;
  max-width: 1200px;
  padding: 40px;
  margin: 0 auto;

  @media screen and (max-width: $tablet) {
    flex-flow: column;
    padding: 15px;
  }
}

.aside {
  flex: 2;
  background-color: white;
  box-shadow: rgba(149, 157, 165, 0.2) 0px 8px 24px;

  @media screen and (max-width: $tablet) {
    order: 2;
  }
}

.main {
  flex: 5;
  padding: 40px;
  background-color: white;
  box-shadow: rgba(149, 157, 165, 0.2) 0px 8px 24px;

  @media screen and (max-width: $tablet) {
    order: 1;
    padding: 40px 10px;
  }
}
