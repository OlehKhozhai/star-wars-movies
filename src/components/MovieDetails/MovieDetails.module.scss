.title {
  display: flex;
  justify-content: space-between;
  font-size: 32px;
  margin-bottom: 20px;
}

.starIcon {
  cursor: pointer;
}
