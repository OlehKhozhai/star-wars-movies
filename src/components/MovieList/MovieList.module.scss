.root {
  padding: 15px 0;
}

.noMovieFound {
  text-align: center;
}
