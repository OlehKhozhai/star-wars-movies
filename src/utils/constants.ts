export enum APIPath {
  Films = "https://swapi.dev/api/films",
}


export enum LSKey {
  FavoriteMovies = "FAVORITE_MOVIES",
}