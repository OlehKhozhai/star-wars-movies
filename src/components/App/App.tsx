import { useEffect, useCallback, useState } from "react";

import MovieList from "components/MovieList/MovieList";
import MovieDetails from "components/MovieDetails/MovieDetails";
import { FavoriteMovies, Movie } from "utils/types";
import {
  getAPIData,
  getFavoriteMoviesConfigFromLS,
  setFavoriteMoviesConfigToLS,
  updateFavoriteMoviesConfig,
  initializeFavoriteMoviesConfigToLS,
} from "utils/helpers";
import { APIPath } from "utils/constants";
import styles from "./App.module.scss";

const App = () => {
  const [movies, setMovies] = useState<Movie[]>([]);
  const [activeMovie, setActiveMovie] = useState<Movie | null>(null);
  const [favoriteMovies, setFavoriteMovies] = useState<FavoriteMovies>(
    getFavoriteMoviesConfigFromLS()
  );

  const handleMovieClick = useCallback((movie: Movie) => {
    setActiveMovie(movie);
  }, []);

  const handleFavoriteIconClick = useCallback((title: string) => {
    const updatedFavoriteMoviesConfig = updateFavoriteMoviesConfig(title);

    setFavoriteMoviesConfigToLS(updatedFavoriteMoviesConfig);
    setFavoriteMovies(updatedFavoriteMoviesConfig);
  }, []);

  useEffect(() => {
    const getMovies = async () => {
      const response = await getAPIData<{ results: Movie[] }>(APIPath.Films);
      if (response?.results?.length > 0) {
        setMovies(response.results);
        setActiveMovie(response.results[0]);
        initializeFavoriteMoviesConfigToLS(response.results);
      }
    };

    getMovies();
  }, []);

  return (
    <div className={styles.root}>
      <aside className={styles.aside}>
        <MovieList
          movies={movies}
          activeMovie={activeMovie}
          favoriteMoviesConfig={favoriteMovies}
          onMovieClick={handleMovieClick}
        />
      </aside>

      {activeMovie && (
        <main className={styles.main}>
          <MovieDetails
            {...activeMovie}
            favoriteMoviesConfig={favoriteMovies}
            onFavoriteIconClick={handleFavoriteIconClick}
          />
        </main>
      )}
    </div>
  );
};

export default App;
