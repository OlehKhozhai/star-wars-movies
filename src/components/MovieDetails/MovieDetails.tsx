import { memo, FC } from "react";

import { StarIcon } from "assets/icons/StarIcon";
import { FavoriteMovies, Movie } from "utils/types";
import styles from "./MovieDetails.module.scss";

type MovieDetailsProps = Movie & {
  favoriteMoviesConfig: FavoriteMovies;
  onFavoriteIconClick: (title: string) => void;
};

const MovieDetails: FC<MovieDetailsProps> = ({
  title,
  opening_crawl,
  favoriteMoviesConfig,
  onFavoriteIconClick,
}) => {
  const handleFavoriteIconClick = () => {
    onFavoriteIconClick(title);
  };

  return (
    <>
      <h1 className={styles.title}>
        {title}
        <span className={styles.starIcon} onClick={handleFavoriteIconClick}>
          <StarIcon fill={favoriteMoviesConfig[title] ? "yellow" : "grey"} />
        </span>
      </h1>

      <p>{opening_crawl}</p>
    </>
  );
};

export default memo(MovieDetails);
