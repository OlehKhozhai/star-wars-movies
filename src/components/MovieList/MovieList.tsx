import { memo, FC } from "react";

import MovieItem from "components/MovieItem/MovieItem";
import { FavoriteMovies, Movie } from "utils/types";
import styles from "./MovieList.module.scss";

type MovieListProps = {
  movies: Movie[];
  activeMovie: Movie | null;
  favoriteMoviesConfig: FavoriteMovies;
  onMovieClick: (movie: Movie) => void;
};

const MovieList: FC<MovieListProps> = ({
  movies,
  activeMovie,
  favoriteMoviesConfig,
  onMovieClick,
}) => {
  return (
    <ul className={styles.root}>
      {movies.length === 0 && (
        <h2 className={styles.noMovieFound}>No movies found</h2>
      )}

      {movies.length > 0 &&
        movies.map((movie) => (
          <MovieItem
            key={movie.title}
            {...movie}
            isActive={movie.title === activeMovie?.title}
            isFavorite={favoriteMoviesConfig[movie.title]}
            onMovieClick={onMovieClick}
          />
        ))}
    </ul>
  );
};

export default memo(MovieList);
