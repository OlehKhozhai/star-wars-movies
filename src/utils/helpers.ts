import { LSKey } from "utils/constants";
import { FavoriteMovies, Movie } from "utils/types";

export const getAPIData = <T>(
  url: string,
  params: RequestInit = {}
): Promise<T> =>
  fetch(url, { ...params })
    .then((data) => data.json())
    .then((data) => data)
    .catch((error) => console.log("--error--", error));

export const initializeFavoriteMoviesConfigToLS = (movies: Movie[]): void => {
  const favoriteMovies = localStorage.getItem(LSKey.FavoriteMovies);

  if (!favoriteMovies) {
    const preparedMovies = movies.reduce<FavoriteMovies>(
      (accumulator, movie) => {
        accumulator[movie.title] = false;

        return accumulator;
      },
      {}
    );

    localStorage.setItem(LSKey.FavoriteMovies, JSON.stringify(preparedMovies));
  }
};

export const getFavoriteMoviesConfigFromLS = (): FavoriteMovies => {
  const favoriteMovies = localStorage.getItem(LSKey.FavoriteMovies);

  return favoriteMovies ? JSON.parse(favoriteMovies) : {};
};

export const setFavoriteMoviesConfigToLS = (config: FavoriteMovies): void => {
  localStorage.setItem(LSKey.FavoriteMovies, JSON.stringify(config));
};

export const updateFavoriteMoviesConfig = (
  title: string
): FavoriteMovies => {
  const favoriteMovies = getFavoriteMoviesConfigFromLS();
  const newFavoriteMovies = {
    ...favoriteMovies,
    [title]: !favoriteMovies[title],
  };

  return newFavoriteMovies;
};
