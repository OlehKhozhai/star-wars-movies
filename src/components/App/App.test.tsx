import React from "react";
import { create, act } from "react-test-renderer";

import { movieMock } from "utils/mock";
import {
  getAPIData,
  getFavoriteMoviesConfigFromLS,
  updateFavoriteMoviesConfig,
} from "utils/helpers";
import { APIPath } from "utils/constants";
import MovieList from "components/MovieList/MovieList";
import MovieDetails from "components/MovieDetails/MovieDetails";
import App from "./App";

jest.mock("./../MovieList/MovieList", () => ({
  __esModule: true,
  default: "<---MovieList--->",
}));
jest.mock("./../MovieDetails/MovieDetails", () => ({
  __esModule: true,
  default: "<---MovieDetails--->",
}));
jest.mock("utils/helpers");

describe("App component", () => {
  beforeAll(() => {
    localStorage.removeItem(APIPath.Films);
  });

  it("should render by default", () => {
    (getFavoriteMoviesConfigFromLS as jest.Mock).mockReturnValue({});
    const component = create(<App />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should render with movies properly", () => {
    (getAPIData as jest.Mock).mockResolvedValue({
      results: [],
    });
    jest
      .spyOn(React, "useState")
      .mockImplementationOnce(() => [[movieMock], jest.fn()])
      .mockImplementationOnce(() => [movieMock, jest.fn()])
      .mockImplementationOnce(() => [{ [movieMock.title]: true }, jest.fn()]);

    const component = create(<App />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should trigger setMovies and setActiveMovie functions with proper arguments", async () => {
    const setMoviesStub = jest.fn();
    const setActiveMovieStub = jest.fn();

    (getAPIData as jest.Mock).mockResolvedValue({
      results: [{ title: "Test" }],
    });
    (getFavoriteMoviesConfigFromLS as jest.Mock).mockReturnValue({});
    jest
      .spyOn(React, "useState")
      .mockImplementationOnce(() => [[movieMock], setMoviesStub])
      .mockImplementationOnce(() => [movieMock, setActiveMovieStub])
      .mockImplementationOnce(() => [{ [movieMock.title]: true }, jest.fn()]);

    await act(async () => {
      create(<App />);
    });

    expect(setMoviesStub).toHaveBeenCalledWith([{ title: "Test" }]);
    expect(setActiveMovieStub).toHaveBeenCalledWith({ title: "Test" });
  });

  it("should trigger handleMovieClick", () => {
    const setActiveMovieStub = jest.fn();
    (getAPIData as jest.Mock).mockResolvedValue({
      results: [],
    });
    jest
      .spyOn(React, "useState")
      .mockImplementationOnce(() => [[movieMock], jest.fn()])
      .mockImplementationOnce(() => [movieMock, setActiveMovieStub])
      .mockImplementationOnce(() => [{ [movieMock.title]: true }, jest.fn()]);

    const component = create(<App />);

    component.root.findByType(MovieList).props.onMovieClick(movieMock);

    expect(setActiveMovieStub).toHaveBeenCalledWith(movieMock);
  });

  it("should trigger handleFavoriteIconClick", () => {
    const setFavoriteMoviesStub = jest.fn();
    (getAPIData as jest.Mock).mockResolvedValue({
      results: [],
    });
    (updateFavoriteMoviesConfig as jest.Mock).mockReturnValue({ Test: true });

    jest
      .spyOn(React, "useState")
      .mockImplementationOnce(() => [[movieMock], jest.fn()])
      .mockImplementationOnce(() => [movieMock, jest.fn()])
      .mockImplementationOnce(() => [
        { [movieMock.title]: true },
        setFavoriteMoviesStub,
      ]);

    const component = create(<App />);

    component.root.findByType(MovieDetails).props.onFavoriteIconClick("Test");

    expect(setFavoriteMoviesStub).toHaveBeenCalledWith({ Test: true });
  });
});
