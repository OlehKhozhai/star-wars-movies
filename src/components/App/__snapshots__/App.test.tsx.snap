// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render by default 1`] = `
<div
  className="root"
>
  <aside
    className="aside"
  >
    <<---MovieList--->
      activeMovie={null}
      favoriteMoviesConfig={Object {}}
      movies={Array []}
      onMovieClick={[Function]}
    />
  </aside>
</div>
`;

exports[`App component should render with movies properly 1`] = `
<div
  className="root"
>
  <aside
    className="aside"
  >
    <<---MovieList--->
      activeMovie={
        Object {
          "created": "2014-12-10T14:23:31.880000Z",
          "director": "George Lucas",
          "episode_id": 4,
          "opening_crawl": "It is a period of civil war.
Rebel spaceships, striking
from a hidden base, have won
their first victory against
the evil Galactic Empire.

During the battle, Rebel
spies managed to steal secret
plans to the Empire's
ultimate weapon, the DEATH
STAR, an armored space
station with enough power
to destroy an entire planet.

Pursued by the Empire's
sinister agents, Princess
Leia races home aboard her
starship, custodian of the
stolen plans that can save her
people and restore
freedom to the galaxy....",
          "producer": "Gary Kurtz, Rick McCallum",
          "release_date": "1977-05-25",
          "title": "A New Hope",
          "url": "https://swapi.dev/api/films/1/",
        }
      }
      favoriteMoviesConfig={
        Object {
          "A New Hope": true,
        }
      }
      movies={
        Array [
          Object {
            "created": "2014-12-10T14:23:31.880000Z",
            "director": "George Lucas",
            "episode_id": 4,
            "opening_crawl": "It is a period of civil war.
Rebel spaceships, striking
from a hidden base, have won
their first victory against
the evil Galactic Empire.

During the battle, Rebel
spies managed to steal secret
plans to the Empire's
ultimate weapon, the DEATH
STAR, an armored space
station with enough power
to destroy an entire planet.

Pursued by the Empire's
sinister agents, Princess
Leia races home aboard her
starship, custodian of the
stolen plans that can save her
people and restore
freedom to the galaxy....",
            "producer": "Gary Kurtz, Rick McCallum",
            "release_date": "1977-05-25",
            "title": "A New Hope",
            "url": "https://swapi.dev/api/films/1/",
          },
        ]
      }
      onMovieClick={[Function]}
    />
  </aside>
  <main
    className="main"
  >
    <<---MovieDetails--->
      created="2014-12-10T14:23:31.880000Z"
      director="George Lucas"
      episode_id={4}
      favoriteMoviesConfig={
        Object {
          "A New Hope": true,
        }
      }
      onFavoriteIconClick={[Function]}
      opening_crawl="It is a period of civil war.
Rebel spaceships, striking
from a hidden base, have won
their first victory against
the evil Galactic Empire.

During the battle, Rebel
spies managed to steal secret
plans to the Empire's
ultimate weapon, the DEATH
STAR, an armored space
station with enough power
to destroy an entire planet.

Pursued by the Empire's
sinister agents, Princess
Leia races home aboard her
starship, custodian of the
stolen plans that can save her
people and restore
freedom to the galaxy...."
      producer="Gary Kurtz, Rick McCallum"
      release_date="1977-05-25"
      title="A New Hope"
      url="https://swapi.dev/api/films/1/"
    />
  </main>
</div>
`;
